### ISSUE TITLE
windows-restart failing to restart

### ISSUE THREAD

**Issue by highlyunavailable** [mitchellh/packer#2613]
I've got a packer config that looks like this:
```
    {
      "type": "powershell",
      "scripts": [
        "{{template_dir}}/scripts/install-netfx3.ps1",
        "{{template_dir}}/scripts/win-updates.ps1"
      ]
    },
    {
      "type": "windows-restart"
    },
    {
      "type": "powershell",
      "scripts": [
        "{{template_dir}}/scripts/win-updates.ps1",
        "{{template_dir}}/scripts/set-bundleconfig.ps1",
        "{{template_dir}}/scripts/set-ec2config.ps1"
      ]
    },
    {
      "type": "windows-restart"
    }
```

I'm using the windows-restart action twice because windows updates don't always apply all of them on the first pass, so I apply all applicable updates, restart, apply again to get more updates, restart, create an AMI.

I'm getting the following non-terminating error on the SECOND windows-restart block:
```
==> amazon-ebs: Restarting Machine
    amazon-ebs: Restart-Computer : Failed to restart the computer WIN-1VVN227ATGQ with the
    amazon-ebs: following error message: A system shutdown is in progress.
    amazon-ebs: At line:1 char:4
    amazon-ebs: + & {Restart-Computer -force }
    amazon-ebs: +    ~~~~~~~~~~~~~~~~~~~~~~~
    amazon-ebs: + CategoryInfo          : OperationStopped: (WIN-1VVN227ATGQ:String) [Rest
    amazon-ebs: art-Computer], InvalidOperationException
    amazon-ebs: + FullyQualifiedErrorId : RestartcomputerFailed,Microsoft.PowerShell.Comma
    amazon-ebs: nds.RestartComputerCommand
    amazon-ebs:
==> amazon-ebs: Waiting for machine to restart...
    amazon-ebs: WIN-1VVN227ATGQ restarted.
==> amazon-ebs: Machine successfully restarted, moving on
```

None of the other scripts I'm using here would restart the machine.

**Comment by highlyunavailable** [comment]
Changed title - the restart actually isn't happening.

Also the docs at https://www.packer.io/docs/provisioners/windows-restart.html are wrong - it says that the default command is `shutdown /r /c "packer restart" /t 5 && net stop winrm` but the actual [provisioner](https://github.com/mitchellh/packer/blob/v0.8.5/provisioner/windows-restart/provisioner.go#L16) uses powershell.

**Comment by cbednarski** [comment]
Thanks for the report. I'm sure there's at least one packer bug here but I'm a bit confused about your description. Your logs seem to indicate that a restart command is pending, but that restart never happens? Or was never invoked? I'm not sure exactly what's happening there.

A few possible interpretations (not mutually exclusive):

- Packer is misinterpreting *something* as a restart event but the box doesn't actually restart
- Packer has a race condition or other weird thing that causes it not to wait for a restart to complete, or to report it has completed too early, or otherwise just not work
- Your scripts are getting stuck somewhere or have already invoked a restart or shutdown, so packer tries to restart but fails
- Windows is being windows
- The docs are busted

If I read the logs in a naive way it looks to me like:

1. Packer tries to restart but can't because Windows is shutting down
2. Packer registers the restart failure as a success and continues merrily along (this is the bug)

Did I follow that correctly?

**Comment by highlyunavailable** [comment]
Yeah, I wasn't 100% certain what's going on here either, but I've isolated it a bit better now.

> If I read the logs in a naive way it looks to me like:
>
> Packer tries to restart but can't because Windows is shutting down
> Packer registers the restart failure as a success and continues merrily along (this is the bug)

You are 100% correct as to the bug: "Packer registers the restart failure as a success and continues merrily along (this is the bug)".

The other thing is Windows being Windows. What's happening is `Restart-Computer` is failing because "there is a shutdown already in progress". However, when I replace the default restart command with `shutdown /a && shutdown /r /t 5 || shutdown /r /t 5` which should try to abort all pending shutdowns and then start a shutdown, or start a shutdown if one is not in progress, it "aborts" successfully, then fails to restart with "A system shutdown is in progress.(1115)". This has something to do with the windows updates I'm doing immediately prior, I'm sure, and will troubleshoot that myself.

**Comment by cbednarski** [comment]
Thanks for clarifying! I *think* for us to fix the bug we should make packer detect the failure correctly so it does not report success for a failed restart. Following this we can wait and retry until the restart succeeds. In your case, though, I think we would end up restarting the box one extra time when it is actually unneeded.

To make this work "correctly" you may need to tweak your scripts a bit.

For linux builds we typically recommend adding a sleep to the script that invokes restart to give restart a chance to happen. Sometimes it takes a while for services to shutdown and such, so the sleep makes packer wait before running the next provisioner. Eventually the box restarts and packer loses its connection, ending that provisioner. At this point the box appears "down" to packer so the next provisioner will fail but Packer then retries until the box is back up and eventually the provisioner succeeds (or errors or times out).

As an alternative to this sleep-down-retry loop, you can also add a [`pause_before` option](https://www.packer.io/docs/templates/provisioners.html) to the subsequent provisioner. This simply tells packer to wait some amount of time before running the next step. The downside here is that it's a bit slower (packer will always wait *at least* that amount of time even if the box is ready sooner) and you have to get the timing right, which involves some guessing. The sleep-down-retry generally works fairly quickly because each step happens as quickly as packer and the VM are ready to go.

**Comment by highlyunavailable** [comment]
I found out if I let it sit long enough it actually does restart, so I'm trying another version of "sleep":

`"restart_command": "shutdown /r /c \"packer restart\" /t 1 || if errorlevel 1115 net stop winrm"`

It'll try to restart, but if it's already in progress, it'll just stop winrm which will return when the box comes back up from the restart.

**Comment by highlyunavailable** [comment]
Ha, yes, it was actually a sleep issue. Ther

**Comment by highlyunavailable** [comment]
Oops. There really should be a sleep by default too.

**Comment by cbednarski** [comment]
> There really should be a sleep by default too.

I'm not sure where I could put this. I think this sleep has to go in your `"{{template_dir}}/scripts/win-updates.ps1"` script.

Each provisioner in packer runs without any knowledge of what came before or is coming after it, so the burden is on you to make sure each provisioner is well-behaved and leaves the machine ready for the next provisioner to run. Unfortunately packer doesn't have any insight into state inside the VM -- like whether it's restarting or has recently restarted or needs to restart -- so we just try to apply each provisioner until it succeeds, fails, or times out.

On the surface this seems simplistic but it allows us to support a wide range of target operating systems since packer doesn't need to know anything about them.

**Comment by highlyunavailable** [comment]
I disagree. I think the [command line](https://github.com/mitchellh/packer/blob/v0.8.5/provisioner/windows-restart/provisioner.go#L16) should read something like 
var DefaultRestartCommand = "powershell \"& {Restart-Computer -force; start-sleep 600}\""`

This is because on Windows, all the "restart" commands (shutdown, restart-computer, etc.) are async and only tell Windows to start shutting down - they will not block until the winrm service stops.

I'm perfectly happy to work around by setting a sleep in my own restart command (which is what I'm doing now) but given how Windows shutdowns work I think it would be better. You say "Unfortunately packer doesn't have any insight into state inside the VM -- like whether it's restarting or has recently restarted or needs to restart -- so we just try to apply each provisioner until it succeeds, fails, or times out." - this is not actually correct, because if I don't use a windows-restart provisioner to execute the restart (e.g. I try to restart the machine by doing it inside my windows update script), the provisioning run actually fails because the network connection to the machine was lost. The only safe way to restart is to use a "restart provisioner" since it can handle reconnecting.

**Comment by cbednarski** [comment]
> var DefaultRestartCommand = "powershell \"& {Restart-Computer -force; start-sleep 600}\""

I see. I think that's reasonable.

> if I don't use a windows-restart provisioner to execute the restart (e.g. I try to restart the machine by doing it inside my windows update script), the provisioning run actually fails because the network connection to the machine was lost.

Ahh, that's interesting. I'll look at this again. After looking at the linux implementation I was under the impression that this was expected and the next provisioner would recover. Windows might just work differently in this case. Thanks again for clarifying!

**Comment by highlyunavailable** [comment]
Actually, this is even more annoying than I thought to work around. If I use `powershell "& {Restart-Computer -force; start-sleep 600}"`, the powershell command exits with error code 1 because the shutdown process aborted it, which means that Packer then fails with `Build 'amazon-ebs' errored: Restart script exited with non-zero exit status: 1`, and windows-restart doesn't support the `valid_exit_codes` value like the shells do.

Further, since the powershell.exe process is actually "canceled" by handling an OS signal, I can't run "exit $lasterrorcode" or whatever. I've had to resort to `powershell "& { Restart-Computer -force; Start-Sleep 600 }" || if errorlevel 1 exit 0` which seems to work properly, but it feels like such a hack, especially since it starts up powershell to do the restart. It looks like it got changed in #2243 but I'm not sure exactly why powershell got pulled into the mix - probably because it "forgets" the error codes. Either way, though, the problem that they had in #2243 still exists with sleeping - the restart script exits with error code 1 because the host process was requested to exit by the OS for shutdown.

It feels like exit code 1 should just be a valid exit code for the windows restart provisioner, and then the normal shutdown.exe with some sort of pause after it can be used - perhaps something like `shutdown /r /t 2 /c \"packer restart\" && pause > nul` (`pause` waits forever for keyboard input).

**Comment by highlyunavailable** [comment]
Annnnd now not even `powershell "& { Restart-Computer -force; Start-Sleep 600 }" || if errorlevel 1 exit 0` only works intermittently. I'm just going to have to use `pause_before` I guess.

**Comment by highlyunavailable** [comment]
More bugs in this provisioner:

`restart_check_command` does not actually respect exit codes. I made a check command that checks the last boot time: `powershell -command "& {$lbt = (Get-WmiObject win32_operatingsystem).LastBootUpTime; if (!(Test-Path C:\ProgramData\lastboot.txt)) {$lbt > C:\ProgramData\lastboot.txt; exit 0} else {if ((get-content C:\ProgramData\lastboot.txt) -eq $lbt) {exit 1} else {remove-item C:\ProgramData\lastboot.txt -force; exit 0}}}"`

And even though it's exiting with a failure (non-zero exit code) it considers that to be a successful check command.

**Comment by pecigonzalo** [comment]
This should partially be fixed by #2568 .
Now to the things your are trying. Doing a sleep after the reboot is going to fail, as soon as windows stops the WinRM service, it terminates the running operation, returning an error to the provisioner.
The second thing is because the GO WinRM library does not properly return exit codes some times from the command that it ran. I submited a PR on that project to help fix that as well.

Adding an arbitrary sleep i think its the wrong thing to do, the arbitrary sleep only work for some cases, now if the reboot takes longer for whatever reason it will not work (EG: Machine is slow, windows updates holding it up, etc). It is better to detect if a reboot is in progress already. 

More so, your previous scripts are causing you the issue, because you are already scheduling a reboot, which you shouldnt outside ```windows-restart```.

Till my PR is merged you have 2 options.
A) Build from my fork and take it for a spin.
B) Add the arbitrary sleep yourself on the next provider using the ```pause_before``` as you said.

Now, i said partially fix at the beginning because its not going to fix the fact that you already have a scheduled reboot. Once the PR is in, you can because of your specific use case change the reboot command to abort the previous one, but i would not make that the default.

Also, powershell was pulled into the mix because the PS rebooter is much better than the CMD alternative, it does not forget the exit codes, what its getting the wrong exit code is your sleep command.

**Comment by highlyunavailable** [comment]
@pecigonzalo I am not scheduling a reboot at all in the other scripts. I determined that the "in progress" reboot was actually the 2nd restart trying to run before the first had finished. The problem I'm running into is this:

1. I run windows update, explicitly NOT rebooting.
3. The restart-windows provisioner sends a reboot, then after a few seconds tries to run a command to see if the machine is back up. Since it hasn't even started rebooting yet, it returns.
4. Since the reboot test command worked, it thinks the machine has rebooted, even though it hasn't even started yet.
5. It runs windows update again, which finishes instantly because there's a reboot required.
6. The second reboot provisioner runs, trying to reboot again, but gets a 1115 error because the first reboot hasn't actually happened yet.

> More so, your previous scripts are causing you the issue, because you are already scheduling a reboot, which you shouldnt outside windows-restart.

This is incorrect, as stated above. It looks like your fix would fix all my problems if you're checking the last boot time.

I'm also trying a short script in `restart_check_command` to see if I can use that to detect the last boot up time:
`powershell -command "& {$ecode = 0; $lbt = (Get-WmiObject win32_operatingsystem).LastBootUpTime; if (!(Test-Path C:\\ProgramData\\lastboot.txt) -or (get-content C:\\ProgramData\\lastboot.txt) -eq $lbt) {$ecode = 1}; $lbt > C:\\ProgramData\\lastboot.txt; exit $ecode}"`

**Comment by pecigonzalo** [comment]
@highlyunavailable Sorry but you are scheduling a reboot, aware or unaware of it, you are doing it. Windows is saying its there, something scheduled the reboot that is not the packer rebooter.

My PR is not checking for boot time. I was planning of adding that as a wrapper in the future, but it complicates stuff and i felt it was not needed initially as the process i implement in the PR works without issues on Vagrant.

The restart_check_command does not actually check if the restart did work, it just checks the machine is up again. But your idea is on point and similar to how the other implantation i mentioned works.

**Comment by highlyunavailable** [comment]
@pecigonzalo No, you didn't read the steps that I wrote above. None of my scripts are scheduling the reboot, it's the windows-restart provisioner that is then not waiting for the reboot to complete before starting the second windows-restart provisioner. Regardless of that, your stuff should fix it - I pulled a copy of your branch and I'm taking a look.

**Comment by pecigonzalo** [comment]
@highlyunavailable My bad, you are right, its the rebooted after the updates the one not finishing before the other rebooter runs. I misread some of the conversation.

**Comment by highlyunavailable** [comment]
Ha ha! I got a working restart_check_command!

`"restart_check_command": "powershell -command \"& {$lbt = (Get-WmiObject win32_operatingsystem).LastBootUpTime; if (!(Test-Path C:\\ProgramData\\lastboot.txt) -or (get-content C:\\ProgramData\\lastboot.txt) -eq $lbt) {Write-Output 'Sleeping for 600 sec'; start-sleep 600}; $lbt > C:\\ProgramData\\lastboot.txt; Write-Output 'Reboot complete'}\""`

**Comment by pecigonzalo** [comment]
Thats great, that gave me an idea on how to implement it on the PR without storing it to a GO Variable. after i fix the error that is not passing ill do that! thanks!

**Comment by highlyunavailable** [comment]
@pecigonzalo I modified it a bit more - a "fast" reboot (that is, it reboots before the check command is run for the first time) always would wait 5 minutes in that scenario. Here's my new full config block:

```
    {
      "type": "windows-restart",
      "restart_command": "powershell \"& {(Get-WmiObject win32_operatingsystem).LastBootUpTime > C:\\ProgramData\\lastboot.txt; Restart-Computer -force}\"",
      "restart_check_command": "powershell -command \"& {if ((get-content C:\\ProgramData\\lastboot.txt) -eq (Get-WmiObject win32_operatingsystem).LastBootUpTime) {Write-Output 'Sleeping for 600 seconds to wait for reboot'; start-sleep 600} else {Write-Output 'Reboot complete'}}\""
    }
```

**Comment by pecigonzalo** [comment]
Does that work? because as far as i see it that would fail the restart_check_command when the server reboots since it kills all running commands. You would never get the Reboot complete. 
You already gave me a good idea on how to implement this inside the rebooter, once i fix the test error ill put it on the PR and let you know.

**Comment by highlyunavailable** [comment]
It sure does work - I tested it last night. The reason is that the reboot check ignores errors and just tries to run the command until it completes (regardless of exit code), which means that it'll start the command, which will start sleeping, then the reboot will happen, which fails the command, and it'll keep retrying the command until a connection is present again at which point it will succeed.

**Comment by markpeek** [comment]
@highlyunavailable can this be closed now that #2568 has been merged and your above comment implies you have your check working?

**Comment by highlyunavailable** [comment]
@markpeek Yes, this can be closed. Thanks!
